### PR TITLE
feat(sql): implement sum function for long256 type

### LINKED ISSUE #1900: Sum on Long256 values
### Is your feature request related to a problem?

I have a `long256` values in my database and was wondering if im able to use `sum` on it to aggregate queries? 

seems this fails for me: 
```sql
CREATE table test (
    a long256,
    b long256
)

INSERT INTO test (a,b) VALUES (0x4f338f76,0x4f338f76);

 SELECT
        sum(a) as sumA, 
        sum(a) as sumB
      FROM test;
```

sum function doesnt work here without casting to double

### Describe the solution you'd like.

sum function to work w long256 values

### Describe alternatives you've considered.

Not sure if you guys could suggest an alternative when working with long256 values

### Additional context.
Screenshot 2022-02-18 at 12 09 39 AM

this here should be `value is double`

### PR DESCRIPTION
**Pull request by eugenels**
Fixes #1900

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 142 / 213 (66.67%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/SumLong256GroupByFunction.java|0|33|00.00%|
|:large_blue_circle:|io/questdb/cairo/map/CompactMapValue.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/cairo/map/FastMapValue.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/SimpleMapValue.java|28|52|53.85%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/SumLong256GroupByFunctionFactory.java|3|4|75.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/math/AddLong256FunctionFactory.java|12|16|75.00%|
|:large_blue_circle:|io/questdb/std/Long256Util.java|14|15|93.33%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/vect/SumLong256VectorAggregateFunction.java|63|67|94.03%|
|:large_blue_circle:|io/questdb/std/Long256Impl.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/vect/GroupByRecordCursorFactory.java|11|11|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Numbers.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 142 / 213 (66.67%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/SumLong256GroupByFunction.java|0|33|00.00%|
|:large_blue_circle:|io/questdb/cairo/map/CompactMapValue.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/cairo/map/FastMapValue.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/SimpleMapValue.java|28|52|53.85%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/SumLong256GroupByFunctionFactory.java|3|4|75.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/math/AddLong256FunctionFactory.java|12|16|75.00%|
|:large_blue_circle:|io/questdb/std/Long256Util.java|14|15|93.33%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/vect/SumLong256VectorAggregateFunction.java|63|67|94.03%|
|:large_blue_circle:|io/questdb/std/Long256Impl.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/vect/GroupByRecordCursorFactory.java|11|11|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Numbers.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 258 / 280 (92.14%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/map/CompactMapValue.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/cairo/map/FastMapValue.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/math/AddLong256FunctionFactory.java|12|16|75.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/SimpleMapValue.java|45|52|86.54%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/SumLong256GroupByFunction.java|30|33|90.91%|
|:large_blue_circle:|io/questdb/std/Long256Util.java|14|15|93.33%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/vect/SumLong256VectorAggregateFunction.java|63|66|95.45%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/SumLong256GroupByFunctionFactory.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/mp/SimpleSpinLock.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/std/Long256Impl.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/vect/GroupByRecordCursorFactory.java|11|11|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Numbers.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGConnectionContext.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlKeywords.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/ExpressionParser.java|30|30|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/eq/EqDoubleFunctionFactory.java|15|15|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 197 / 219 (89.95%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/map/CompactMapValue.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/cairo/map/FastMapValue.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/math/AddLong256FunctionFactory.java|12|16|75.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/SimpleMapValue.java|45|52|86.54%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/SumLong256GroupByFunction.java|30|33|90.91%|
|:large_blue_circle:|io/questdb/std/Long256Util.java|14|15|93.33%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/vect/SumLong256VectorAggregateFunction.java|63|66|95.45%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/SumLong256GroupByFunctionFactory.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/mp/SimpleSpinLock.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/std/Long256Impl.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/vect/GroupByRecordCursorFactory.java|11|11|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Numbers.java|1|1|100.00%|

File: core/src/main/java/io/questdb/griffin/engine/groupby/vect/GroupByRecordCursorFactory.java [hunk 1]
```diff
@@ -438,17 +438,25 @@ public long getLong(int col) {
...
 
             @Override
             public Long256 getLong256A(int col) {
-                return null;
+                final long offset = getValueOffset(col);
+                final long l0 = Unsafe.getUnsafe().getLong(offset);
+                final long l1 = Unsafe.getUnsafe().getLong(offset + Long.BYTES);
+                final long l2 = Unsafe.getUnsafe().getLong(offset + 2 * Long.BYTES);
+                final long l3 = Unsafe.getUnsafe().getLong(offset + 3 + Long.BYTES);
+                Long256Impl res = new Long256Impl();
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/groupby/vect/GroupByRecordCursorFactory.java:452]: We should not be allocating on execution path. Typical impl is to have two `Long256Impl` per record, e.g. A and B. This also requires a test, where result of this record implementation can be ordered correctly. E.g. `order by long256`

File: core/src/main/java/io/questdb/griffin/engine/groupby/vect/GroupByRecordCursorFactory.java [hunk 2]
```diff
@@ -438,17 +438,25 @@ public long getLong(int col) {
...
+                res.setAll(l0, l1, l2, l3);
+                return res;
             }
 
             @Override
             public Long256 getLong256B(int col) {
-                return null;
+                return getLong256A(col);
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/groupby/vect/GroupByRecordCursorFactory.java:459]: this should be using B instance.

File: core/src/main/java/io/questdb/griffin/engine/groupby/vect/SumLong256VectorAggregateFunction.java [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+    @Override
+    public void getLong256(Record rec, CharSink sink) {
+        Long256Impl v = (Long256Impl) getLong256A(rec);
+        v.toSink(sink);
+    }
+
+    @Override
+    public Long256 getLong256A(Record rec) {
+        Long256Impl res = new Long256Impl();
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/groupby/vect/SumLong256VectorAggregateFunction.java:135]: same here

File: core/src/main/java/io/questdb/griffin/engine/groupby/vect/SumLong256VectorAggregateFunction.java [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+        return Long256Impl.NULL_LONG256;
+    }
+
+    @Override
+    public Long256 getLong256B(Record rec) {
+        return getLong256A(rec);
+    }
+
+    private Long256Impl sumLong256(long address, long count) {
+        Long256Impl sum = new Long256Impl();
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/groupby/vect/SumLong256VectorAggregateFunction.java:149]: same

File: core/src/main/java/io/questdb/std/Long256Impl.java [hunk 5]
```diff
@@ -97,12 +99,23 @@ public void toSink(CharSink sink) {
         Numbers.appendLong256(l0, l1, l2, l3, sink);
     }
 
+    public static Long256Impl add(final Long256 x, final Long256 y) {
+        if (x.equals(Long256Impl.NULL_LONG256) || y.equals(Long256Impl.NULL_LONG256)) {
+            return Long256Impl.NULL_LONG256;
+        }
+        Long256Impl sum = new Long256Impl();
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/std/Long256Impl.java:106]: should be allocation free

**Review by bluestreak01 — CHANGES_REQUESTED** [review]

File: core/src/main/java/io/questdb/griffin/engine/groupby/vect/GroupByRecordCursorFactory.java (hunk 2 above)
Comment (eugenels) [reply core/src/main/java/io/questdb/griffin/engine/groupby/vect/GroupByRecordCursorFactory.java:459]: Can I use Long256Impl.NULL_LONG256 for this "B" case?

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 424 / 512 (82.81%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/map/CompactMapValue.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/cairo/map/FastMapValue.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/log/SyncLogger.java|34|93|36.56%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/math/AddLong256FunctionFactory.java|12|16|75.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/SimpleMapValue.java|45|52|86.54%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/SumLong256GroupByFunction.java|30|33|90.91%|
|:large_blue_circle:|io/questdb/cairo/RecoverVarIndex.java|42|46|91.30%|
|:large_blue_circle:|io/questdb/std/Long256Util.java|14|15|93.33%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/vect/SumLong256VectorAggregateFunction.java|67|70|95.71%|
|:large_blue_circle:|io/questdb/cairo/RebuildIndex.java|35|36|97.22%|
|:large_blue_circle:|io/questdb/cairo/RebuildColumnBase.java|92|94|97.87%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/SumLong256GroupByFunctionFactory.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/log/LogFactory.java|19|19|100.00%|
|:large_blue_circle:|io/questdb/mp/SimpleSpinLock.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/std/Long256Impl.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/vect/GroupByRecordCursorFactory.java|13|13|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Numbers.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 202 / 224 (90.18%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/map/CompactMapValue.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/cairo/map/FastMapValue.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/math/AddLong256FunctionFactory.java|12|16|75.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/SimpleMapValue.java|45|52|86.54%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/SumLong256GroupByFunction.java|30|33|90.91%|
|:large_blue_circle:|io/questdb/std/Long256Util.java|14|15|93.33%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/vect/SumLong256VectorAggregateFunction.java|67|70|95.71%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/SumLong256GroupByFunctionFactory.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/mp/SimpleSpinLock.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/std/Long256Impl.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/vect/GroupByRecordCursorFactory.java|13|13|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Numbers.java|1|1|100.00%|

**Pull request merged**